### DISCUSSION TITLE
Does Quasar Support pnpm Monorepo Setup?

### DISCUSSION THREAD

**Discussion by pinghe** [quasarframework/quasar#17515] [General - Components / Directives / etc] [answered]
Hello everyone,

I am in the process of setting up a monorepo project with pnpm as the package manager. Within this monorepo, I have a main application developed with Quasar, and also within the packages directory, there is another application developed with Quasar.

Currently, I am encountering an issue where it seems that the monorepo structure is not supported by the Quasar CLI. Specifically, I am facing the following problems:

When trying to run quasar dev or quasar build commands, I am receiving an error indicating that the Quasar CLI does not allow execution within a monorepo structure.

I have attempted to run the Quasar CLI separately for each application from the root of the monorepo, but it still doesn't work as expected.

I would like to inquire if anyone has come across a similar issue, or if there are any recommended solutions? Is there official support or a workaround to achieve this structure?

Any advice or guidance would be greatly appreciated.

Thank you!

**Comment by pinghe** [comment]
````
root
    package.json
    - packages
             - mainApp(quasar)
                     package.json  dependencies- "childApp": "workspace:*",  // failed
             - childApp

current:

root(main quasar)
      quasar.config.js
      package.json  dependencies- "childApp": "workspace:*",  // failed
      - packages
             - childApp(quasar)  
  ````

**Comment by rstoenescu** [comment, marked as answer]
It certainly does work with monorepos. I assume that you are using PNPM with it? Make sure that you have this in your `/.npmrc`:

```
# pnpm-related options
shamefully-hoist=true
strict-peer-dependencies=false
# to get the latest compatible packages when creating the project https://github.com/pnpm/pnpm/issues/6463
resolution-mode=highest
```

**Comment by lugrinder** [comment]
Hi, I have a monorepo using pnpm (quasar front end and expressjs backend), and mainly works, but with some issues. When I run `quasar upgrade`  or `pnpm quasar upgrade` in the quasar repo, I get an the next error:

```console
npm error code ENOWORKSPACES.
npm error This command does not support workspaces.

```
How can I upgrade quasar with pnpm in a monorepo? It seems that is trying to use npm for the upgrade instead of pnpm.

**Reply by yusufkandemir** [reply]
Do you run it inside the Quasar project or within the monorepo root?

**Reply by lugrinder** [reply]
Inside Quasar project. Running in monorepo root gets another error (ERR_INVALID_ARG_TYPE)

**Reply by yusufkandemir** [reply]
What does `quasar --version` and `pnpm quasar --version` give? Does `pnpm quasar info` run, if yes please share the output.

Do you have `package-lock.json` in the project or within the monorepo root? Also, do you have `pnpm-lock.yaml` present in the monorepo root?

**Reply by lugrinder** [reply]
```shell
$ quasar --version
@quasar/app-vite 1.10.2 (@quasar/cli 2.4.1)
```

```shell
$ pnpm quasar --version
@quasar/app-vite 1.10.2 (@quasar/cli 2.4.1)
```

```shell
$ pnpm quasar info

Operating System - Windows_NT(10.0.19045) - win32/x64
NodeJs - 20.13.1

Global packages
  NPM - 10.8.3
  yarn - Not installed
  @quasar/cli - 2.4.1
  @quasar/icongenie - 4.0.0
  cordova - Not installed

Important local packages
  quasar - 2.17.2 -- Build high-performance VueJS user interfaces (SPA, PWA, SSR, Mobile and Desktop) in record time
  @quasar/app-vite - 1.10.2 -- Quasar Framework App CLI with Vite
  @quasar/extras - 1.16.13 -- Quasar Framework fonts, icons and animations
  eslint-plugin-quasar - Not installed
  vue - 3.5.12 -- The progressive JavaScript framework for building modern web UI.
  vue-router - 4.4.5
  pinia - 2.2.6 -- Intuitive, type safe and flexible Store for Vue
  vuex - Not installed
  vite - 2.9.18 -- Native-ESM powered web dev build tool
  eslint - 8.57.1 -- An AST-based pattern checker for JavaScript.
  electron - Not installed
  electron-packager - Not installed
  @electron/packager - Not installed
  electron-builder - Not installed
  register-service-worker - 1.7.2 -- Script for registering service worker, with hooks
  @capacitor/core - Not installed
  @capacitor/cli - Not installed
  @capacitor/android - Not installed
  @capacitor/ios - Not installed

Quasar App Extensions
  *None installed*
```

Prior to `pnpm` migration, I've deleted all `node_modules` and `package-lock.json`, from all monorepo packages and root.
I've installed all packages from monorepo root (`pnpm install`) and only one `pnpm-lock.yaml` file, present at monorepo root. After migration and minor fixes, all packages (backend and quasar) builds without any error. After first compile without any errors, I've deleted all `node_modules` again to ensure a clean installation and rerun `pnpm install`.

**Reply by yusufkandemir** [reply]
So, with this exact stuff and place that you are running the commands in, please share the exact output of `pnpm quasar upgrade`.

**Reply by lugrinder** [reply]
```shell
$ pnpm quasar upgrade

npm warn config ignoring workspace config at D:\Sites\app\packages\quasar/.npmrc
npm error code ENOWORKSPACES
npm error This command does not support workspaces.
npm error A complete log of this run can be found in: C:\Users\User\AppData\Local\npm-cache\_logs\2024-11-12T15_35_59_404Z-debug-0.log
 Global Quasar CLI • Gathering information from the NPM registry (https://registry.npmjs.org/)...

 Global Quasar CLI • Congrats! All Quasar packages are up to date (according to https://registry.npmjs.org/).
```

The complete log at C:\Users\User\AppData\Local\npm-cache\_logs\2024-11-12T15_35_59_404Z-debug-0.log:

```shell
0 verbose cli C:\Program Files\nodejs\node.exe C:\Program Files\nodejs\node_modules\npm\bin\npm-cli.js
1 info using npm@10.8.3
2 info using node@v20.13.1
3 silly config load:file:C:\Users\User\AppData\Roaming\nvm\v20.13.1\node_modules\npm\npmrc
4 warn config ignoring workspace config at D:\Sites\app\packages\quasar/.npmrc
5 info config found workspace root at D:\Sites\app
6 silly config load:file:D:\Sites\app\.npmrc
7 silly config load:file:C:\Users\User\.npmrc
8 silly config load:file:C:\Program Files\nodejs\etc\npmrc
9 verbose title npm config get registry
10 verbose argv "config" "get" "registry"
11 verbose logfile logs-max:10 dir:C:\Users\User\AppData\Local\npm-cache\_logs\2024-11-12T15_35_59_404Z-
12 verbose logfile C:\Users\User\AppData\Local\npm-cache\_logs\2024-11-12T15_35_59_404Z-debug-0.log
13 silly logfile start cleaning logs, removing 1 files
14 verbose stack Error: This command does not support workspaces.
14 verbose stack     at #exec (C:\Users\User\AppData\Roaming\nvm\v20.13.1\node_modules\npm\lib\npm.js:243:29)
14 verbose stack     at Npm.exec (C:\Users\User\AppData\Roaming\nvm\v20.13.1\node_modules\npm\lib\npm.js:207:25)
14 verbose stack     at module.exports (C:\Users\User\AppData\Roaming\nvm\v20.13.1\node_modules\npm\lib\cli\entry.js:65:29)
15 error code ENOWORKSPACES
16 error This command does not support workspaces.
17 verbose cwd D:\Sites\app\packages\quasar
18 verbose os Windows_NT 10.0.19045
19 verbose node v20.13.1
20 verbose npm  v10.8.3
21 verbose exit 1
22 verbose code 1
23 error A complete log of this run can be found in: C:\Users\User\AppData\Local\npm-cache\_logs\2024-11-12T15_35_59_404Z-debug-0.log

```

**Reply by yusufkandemir** [reply]
Can you share the contents of `D:\Sites\app\packages\quasar/.npmrc`?
Btw, unless you have project-specific options, you can put it in the workspace root. I believe some options are only effective if inside the workspace root.

**Reply by TobyMosque** [reply]
without -i

![image](https://github.com/user-attachments/assets/e20dccc7-05c7-4386-beb3-f47179bab4ce)

with -i

![image](https://github.com/user-attachments/assets/10ef2f88-561a-4898-857c-b5afd9b67c68)

https://github.com/TobyMosque/ws-quasar-nest-20241112
in order to start the project, run `pnpm monolith`, so access `localhost:9100` and/or `localhost:9100/api/docs`

**Reply by lugrinder** [reply]
> Can you share the contents of `D:\Sites\app\packages\quasar/.npmrc`?

```
# pnpm-related options
shamefully-hoist=true
strict-peer-dependencies=false
# to get the latest compatible packages when creating the project https://github.com/pnpm/pnpm/issues/6463
resolution-mode=highest
```
> Btw, unless you have project-specific options, you can put it in the workspace root. I believe some options are only effective if inside the workspace root.

In a first step, I've copied `.npmrc` file to the root to workspace, same results.
In a second step, I've deleted `.npmrc` file (keeping in it the root) from quasar package, same results.

**Reply by yusufkandemir** [reply]
If you run `npm config get registry` in the monorepo root or in the project folder, does it give an error or run fine?

**Reply by lugrinder** [reply]
> If you run `npm config get registry` in the monorepo root or in the project folder, does it give an error or run fine?

In monorepo root:
```shell
$ npm config get registry
https://registry.npmjs.org/
```

In quasar folder:
```shell
$ npm config get registry
npm warn config ignoring workspace config at D:\Sites\app\packages\quasar/.npmrc
npm error code ENOWORKSPACES
npm error This command does not support workspaces.
npm error A complete log of this run can be found in: C:\Users\User\AppData\Local\npm-cache\_logs\2024-11-13T07_31_00_098Z-debug-0.log
```

The log file contains the same as previously reported.

**Reply by yusufkandemir** [reply]
Now I pretty much highly suspect you have npm workspace configuration like `package.json > workspaces` somewhere. Can you verify if that's the case?

**Reply by yusufkandemir** [reply]
Btw, due to https://github.com/npm/cli/issues/6099, `npm config` command doesn't work in workspaces. So, that is why this error is happening.

**Reply by lugrinder** [reply]
> Now I pretty much highly suspect you have npm workspace configuration like `package.json > workspaces` somewhere. Can you verify if that's the case?

Yes, that's the problem. I've migrated the two apps to a monorepo, and at first attempt, I've done with `npm` adding workspace to the monorepo package. Thank you very much for your help.

I've asked for support in #15941 about importing from backend into quasar package. Can you help me about this? This is the main reason for the monorepo creation, keep types (I'm using typescript) in sync from backend and quasar.

**Reply by yusufkandemir** [reply]
For direct usage of the server package in the client, you need to use deep imports like `@my-scope/server-pkg/src/something/whatever`. If the file contains decorators, or node-specific lib code, you would need to split the file into a "type-only" file and the rest. So, make sure there are only types or code that can run on the client.

For a relatively more complex but better solution, you would need to create a shared package in your monorepo. Then, you'd use that package in the client and the server.

**Reply by lugrinder** [reply]
> For direct usage of the server package in the client, you need to use deep imports like `@my-scope/server-pkg/src/something/whatever`. If the file contains decorators, or node-specific lib code, you would need to split the file into a "type-only" file and the rest. So, make sure there are only types or code that can run on the client.
> 

I've done that, was my first action, but don't work. But I've solved (accidentally) the problem with a simple solution (added yesterday for other questions): adding the scope to the `build.alias` of `quasar.conf.js`.

> For a relatively more complex but better solution, you would need to create a shared package in your monorepo. Then, you'd use that package in the client and the server. Now builds without any error.

This is my intention.

Thank you very much for your help.

**Comment by lugrinder** [comment]
@yusufkandemir I have a minor issue with install/build in the monorepo that I hope that you can help. If I install in the root with `pnpm install`, the backend builds without any problem, but the quasar package shows errors on missing dependencies:

```
[vite]: Rollup failed to resolve import "register-service-worker" from "src-pwa/register-service-worker.ts".
This is most likely unintended because it can break your application at runtime.
If you do want to externalize this module explicitly add it to
`build.rollupOptions.external`
```

But If I install inside the quasar package, builds now works without any problem...

**Reply by yusufkandemir** [reply]
Make sure the monorepo is set up properly. You can run `pnpm m ls --depth -1` to see if the Quasar project is properly included in the monorepo. Move the `.npmrc` file from the Quasar project to the monorepo root. Here is an example `.npmrc` file that should be in the monorepo root:
```
# pnpm-related options
shell-emulator=true
shamefully-hoist=true
strict-peer-dependencies=false
```

**Reply by yusufkandemir** [reply]
Also, `register-service-worker` is a bit suspicious as it used to be provided from the app-* packages, but now should be installed into the project by the user. So, make sure it is inside your `package.json`.

**Reply by lugrinder** [reply]
Moving `.npmrc` to root was sufficient. Thanks.

**Comment by RobsonMi** [comment]
Sorry to hijack a little but I have a very similar problem with Electron Quasar app. I need to share some components between 2 apps therefore I have created a monorepo using pnpm as package manager. Unfortunately during `Installing UnPackaged folder production dependencies...` build step pnpm asks to completely overwrite the root node_modules instead of creating a temporary node_modules inside the UnPackaged directory. I used the `.npmrc` file in the root of monorepo with following contents:
```
# pnpm-related options
shamefully-hoist=true
strict-peer-dependencies=false
# to get the latest compatible packages when creating the project https://github.com/pnpm/pnpm/issues/6463
resolution-mode=highest
```

Electron app works flawlessly when started using `quasar dev -m electron`. Am I missing some important setup step to allow for sucessful build of Electron Quasar app in monorepo?

**Reply by RobsonMi** [reply]
I did discover that probably using `--ignore-workspace` flag in `pnpm install --prod --ignore-workspace` would solve my issues but sadly adding `npmArgs: '--ignore-workspace'` in `quasar.config.js` builder section doesn't work like I expected.

**Reply by yusufkandemir** [reply]
Here's the solution:
```
electron: {
  // Quasar default: `[ 'i', '--prod' ]`
  unPackagedInstallParams: [
    "i",
    "--prod",
    // pnpm picks up the root `prepare` script and fails running it as devDependencies are not installed
    // See https://github.com/pnpm/pnpm/issues/7068
    "--ignore-scripts",
    // Even if we don't declare it as a workspace, pnpm will still act like it is, but also not really
    // So, to avoid a huge array of problems, we have to ignore the workspace
    // See https://github.com/orgs/pnpm/discussions/4735#discussioncomment-2770033
    "--ignore-workspace",
  ],
}
```
We'll try to standardize this and put it into Quasar CLI in the future.

**Reply by RobsonMi** [reply]
Awesome! I did find the `unPackagedInstallParams` on my own but removed `postinstall` script to avoid issues. Now I can add it back. Thank you ;)

**Reply by RobsonMi** [reply]
@yusufkandemir if I may ask you one more question. I managed to get Quasar with Electron going in PNPM Workspace but soon discovered that dependencies used by electron-builder have some crazy low versions. Say in my package.json I have `"got": "^14.4.7"` but in UnPackaged package.json it's `"got": "11.0.2"`, same for few other dependencies. This basically breaks the application. Is there any way to avoid this?

**Reply by yusufkandemir** [reply]
This is the relevant piece of code:
https://github.com/quasarframework/quasar/blob/dev/app-vite/lib/utils/get-fixed-deps.js

It resolves the actual used dependencies and uses their version. We would need more information to say whether you are experiencing an edge case, or if it's just that your package.json does not match your lock file or similar.
